chain/processor.go

```diff
@@ -16,15 +22,221 @@ import (
+		feeKeyStr:       parentFeeManager.Bytes(),
+	})
+	if err := tsv.Insert(ctx, heightKey, binary.BigEndian.AppendUint64(nil, b.Hght)); err != nil {
+		return nil, nil, nil, err
+	}
+	if err := tsv.Insert(ctx, timestampKey, binary.BigEndian.AppendUint64(nil, uint64(b.Tmstmp))); err != nil {
+		return nil, nil, nil, err
+	}
+	if err := tsv.Insert(ctx, feeKey, feeManager.Bytes()); err != nil {
+		return nil, nil, nil, err
+	}
+	tsv.Commit()
+
+	// Compare state root
+	//
+	// Because fee bytes are not recorded in state, it is sufficient to check the state root
+	// to verify all fee calculations were correct.
+	_, rspan := e.vm.Tracer().Start(ctx, "StatefulBlock.Verify.WaitRoot")
+	start := time.Now()
+	computedRoot, err := parentView.GetMerkleRoot(ctx)
+	rspan.End()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	e.vm.RecordWaitRoot(time.Since(start))
+	if b.StateRoot != computedRoot {
+		return nil, nil, nil, fmt.Errorf(
+			"%w: expected=%s found=%s",
+			ErrStateRootMismatch,
+			computedRoot,
+			b.StateRoot,
+		)
+	}
+
+	// Ensure signatures are verified
+	_, sspan := e.vm.Tracer().Start(ctx, "StatefulBlock.Verify.WaitSignatures")
+	start = time.Now()
+	err = b.sigJob.Wait()
+	sspan.End()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	e.vm.RecordWaitSignatures(time.Since(start))
+
+	// Get view from [tstate] after processing all state transitions
+	e.vm.RecordStateChanges(ts.PendingChanges())
+	e.vm.RecordStateOperations(ts.OpIndex())
+	view, err := ts.ExportMerkleDBView(ctx, e.vm.Tracer(), parentView)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// Kickoff root generation
+	go func() {
+		start := time.Now()
+		root, err := view.GetMerkleRoot(ctx)
+		if err != nil {
+			log.Error("merkle root generation failed", zap.Error(err))
+			return
+		}
+		log.Info("merkle root generated",
+			zap.Uint64("height", b.Hght),
+			zap.Stringer("blkID", b.ID()),
+			zap.Stringer("root", root),
+		)
+		e.vm.RecordRootCalculated(time.Since(start))
+	}()
+	return results, feeManager, view, nil
+}
+
+// verifyExpiryReplayProtection handles expiry replay protection
+//
+// Replay protection confirms a transaction has not been included within the
+// past validity window.
+// Before the node is ready (we have synced a validity window of blocks), this
+// function may return an error when other nodes see the block as valid.
+//
+// If a block is already accepted, its transactions have already been added
+// to the VM's seen emap and calling [IsRepeat] will return a non-zero value
+// when it should already be considered valid, so we skip this step here.
+func (*Executor) verifyExpiryReplayProtection(ctx context.Context, b *StatefulBlock, rules Rules, vctx VerifyContext) error {
+	if b.accepted {
+		return nil
+	}
+
+	oldestAllowed := b.Tmstmp - rules.GetValidityWindow()
+	if oldestAllowed < 0 {
+		// Can occur if verifying genesis
+		oldestAllowed = 0
+	}
+	dup, err := vctx.IsRepeat(ctx, oldestAllowed, b.Txs, set.NewBits(), true)
```

Comment:
This isn't super-efficient, as we're using `vm.seen.Contains` instead of the more appropriate `vm.seen.Any`.
Not an issue for this PR, though.